src/apollo-fetch.ts

```diff
@@ -0,0 +1,162 @@
+import {
+  FetchResult,
+  RequestAndOptions,
+  ResponseAndOptions,
+  AfterwareInterface,
+  MiddlewareInterface,
+  FetchOptions,
+  ApolloFetch,
+  ParsedResponse,
+  GraphQLRequest,
+} from './types';
+import 'isomorphic-fetch';
+
+export function createApolloFetch(params: FetchOptions = {}): ApolloFetch {
+  const {uri, customFetch} = params;
+
+  const _uri = uri || '/graphql';
+  const _middlewares = [];
+  const _afterwares = [];
+
+  const applyMiddlewares = (requestAndOptions: RequestAndOptions): Promise<RequestAndOptions> => {
+    return new Promise((resolve, reject) => {
+      const { request, options } = requestAndOptions;
+      const queue = (funcs: MiddlewareInterface[], scope: any) => {
+        const next = () => {
+          if (funcs.length > 0) {
+            const f = funcs.shift();
+            if (f) {
+              f.applyMiddleware.apply(scope, [{ request, options }, next]);
+            }
+          } else {
+            resolve({
+              request,
+              options,
+            });
+          }
+        };
+        next();
+      };
+
+      queue([..._middlewares], this);
+    });
+  };
+
+
+  const applyAfterwares = ({response, options}: ResponseAndOptions): Promise<ResponseAndOptions> => {
+    return new Promise((resolve, reject) => {
+      // Declare responseObject so that afterware can mutate it.
+      const responseObject = {response, options};
+      const queue = (funcs: AfterwareInterface[], scope: any) => {
+        const next = () => {
+          if (funcs.length > 0) {
+            const f = funcs.shift();
+            if (f) {
+              f.applyAfterware.apply(scope, [responseObject, next]);
+            }
+          } else {
+            resolve(responseObject);
+          }
+        };
+        next();
+      };
+
+      // iterate through afterwares using next callback
+      queue([..._afterwares], this);
+    });
+  };
+
+  const callFetch = ({request, options}) => {
+    const opts = {
+      body: JSON.stringify(request),
+      method: 'POST',
+      ...options,
+      headers: {
+        Accept: '*/*',
+        'Content-Type': 'application/json',
+        ...(options.headers as { [headerName: string]: string }),
```

Comment:
The type of `options.headers` could be defined with RequestAndOptions or something.